File: frontend/src/core/auth/modules.ts
import type { AuthEntity } from "./store";

export type EntityModuleFlag =
  | "enable_pqrs"
  | "enable_users_admin"
  | "enable_reports_pdf"
  | "enable_ai_reports"
  | "enable_planes_institucionales"
  | "enable_contratacion"
  | "enable_pdm"
  | "enable_asistencia"
  | "enable_correspondencia"
  | "enable_presupuesto";

export const MODULE_FLAG_TO_KEY: Record<EntityModuleFlag, string> = {
  enable_pqrs: "pqrs",
  enable_users_admin: "users_admin",
  enable_reports_pdf: "reports_pdf",
  enable_ai_reports: "ai_reports",
  enable_planes_institucionales: "planes_institucionales",
  enable_contratacion: "contratacion",
  enable_pdm: "pdm",
  enable_asistencia: "asistencia",
  enable_correspondencia: "correspondencia",
  enable_presupuesto: "presupuesto",
};

export function isModuleEnabled(entity: AuthEntity, module: EntityModuleFlag): boolean {
  return Boolean(entity[module]);
}

/** Alineado con `is_platform_superadmin` del backend. */
export function isPlatformSuperadmin(
  user: { roles?: string[]; role?: string; is_superuser?: boolean } | null | undefined,
): boolean {
  if (!user) return false;
  if (user.is_superuser) return true;
  return user.roles?.includes("superadmin") || user.role === "superadmin";
}

export function primaryRole(user: { roles?: string[]; role?: string } | null | undefined): string {
  if (!user) return "";
  if (user.role) return user.role;
  const priority = ["superadmin", "admin", "secretario", "ciudadano"];
  for (const r of priority) {
    if (user.roles?.includes(r)) return r;
  }
  return user.roles?.[0] || "";
}

/** Secretario: lista vacía = ningún módulo (asignación explícita). Admin/otros: vacía = todos los de la entidad. */
export function isUserModuleEnabled(
  user: { enabled_modules?: string[]; roles?: string[]; role?: string } | null | undefined,
  module: EntityModuleFlag,
): boolean {
  if (!user) return false;
  const enabled = user.enabled_modules ?? [];
  const role = primaryRole(user);
  if (role === "secretario") {
    if (enabled.length === 0) return false;
    return enabled.includes(MODULE_FLAG_TO_KEY[module]);
  }
  if (enabled.length === 0) return true;
  return enabled.includes(MODULE_FLAG_TO_KEY[module]);
}

File: backend/apps/common/modules.py
"""Validación de módulos por entidad y usuario."""

from __future__ import annotations

from rest_framework.exceptions import PermissionDenied

from apps.common.roles import primary_role

ENTITY_MODULE_FLAGS = {
    "pqrs": "enable_pqrs",
    "users_admin": "enable_users_admin",
    "reports_pdf": "enable_reports_pdf",
    "ai_reports": "enable_ai_reports",
    "planes_institucionales": "enable_planes_institucionales",
    "contratacion": "enable_contratacion",
    "pdm": "enable_pdm",
    "asistencia": "enable_asistencia",
    "correspondencia": "enable_correspondencia",
    "presupuesto": "enable_presupuesto",
}


def entity_has_module(entity, module_key: str) -> bool:
    if entity is None:
        return False
    flag = ENTITY_MODULE_FLAGS.get(module_key)
    if not flag:
        return False
    return bool(getattr(entity, flag, False))


def user_has_module(user, module_key: str) -> bool:
    entity = getattr(user, "entity", None)
    if not entity_has_module(entity, module_key):
        return False
    user_modules = getattr(user, "enabled_modules", None) or []
    role = primary_role(user)
    if role == "secretario":
        if not user_modules:
            return False
        return module_key in user_modules
    if user_modules:
        return module_key in user_modules
    return True


def require_user_module(user, module_key: str, *, message: str | None = None) -> None:
    if not user_has_module(user, module_key):
        raise PermissionDenied(message or f"El módulo '{module_key}' no está habilitado.")
